Generator/CompiledInterceptor.php

```diff
@@ -619,16 +619,54 @@ private function getCompiledMethodInfo(\ReflectionMethod $method, $config)
             $body[] = "}";
         }
 
+        $returnTypeValue = null;
         $returnType = $method->getReturnType();
-        $returnTypeValue = $returnType
-            ? ($returnType->allowsNull() ? '?' : '') . $returnType->getName()
-            : null;
-        if ($returnTypeValue === 'self') {
-            $returnTypeValue = $method->getDeclaringClass()->getName();
+        $isUnionReturnType = ($method->hasReturnType() && $method->getReturnType() instanceof \ReflectionUnionType);
+        if ($isUnionReturnType) {
+            foreach ($method->getReturnType()->getTypes() as $methodReturnTypesItem) {
+                $methodReturnTypeValue = $methodReturnTypesItem
+                    ? ($methodReturnTypesItem->getName() !== 'null' && $methodReturnTypesItem->allowsNull() ? '?' : '') . $methodReturnTypesItem->getName()
```

Comment:
Added condition `$methodReturnTypesItem->getName() !== 'null'` as without it, we get an error
```
Provided type "?null" is not recognized as a valid expression: it must match "/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*(\\[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*)*$/" or be one of the built-in types (1, 2, 3  
  , 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14)                                                                                                                                                                            
  Class Smile\StoreLocator\Controller\Store\ContactPost\Interceptor generation error: The requested class did not generate properly, because the 'generated' directory permission is read-only. If --- after runnin  
  g the 'bin/magento setup:di:compile' CLI command when the 'generated' directory permission is set to write --- the requested class did not generate properly, then you must add the generated class object to the  
   signature of the related construct method, only. 
```

The reason is that, if the return type is "null", it also satisfies "allowsNull" condition, and we get "?null" as a result of value mapping,